### DISCUSSION TITLE
How to keep the connection open when returned to pool?

### DISCUSSION THREAD

**Discussion by sahil-khanna-instacart** [sqlalchemy/sqlalchemy#6210] [Usage Questions]
**Describe your question**
I am trying to use SQLalchemy engine to create database connections and keep it as python class attributes to use it repetitively in consecutive calls. I am using following code to create an engine instance and connection attribute.

Every time, I am calling `func_name` method, it is recreating the connection with the database. I was expecting it to create a connection with DB on first call and reuse the connection from the pool.

Is there a way to return the connection to the Pool without closing it?

```
def __init__(self):
    self.engine = sqlalchemy.create_engine(url, **kwargs).execution_options(
            autocommit=False
        )

@contextlib.contextmanager
def connection(self):
    with self.engine.begin() as connection:
        yield connection
        
# Usage
def func_name(...):
    with self.connection() as connection:
         return connection.execute(sql, params).fetchall()
```

**Example - please use the [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) guidelines if possible**

**Complete stack trace, if applicable**

**Versions**
 - OS: Darwin
 - Python: 3.9
 - SQLAlchemy: 1.4.3
 - Database: Docker image: postgres:13-alpine
 - DBAPI:

**Additional context**

**Useful links**
- [Get Support](https://www.sqlalchemy.org/support.html) on the website
- The [documentation](https://docs.sqlalchemy.org/en/latest/) website
- The [UsageRecipes](https://github.com/sqlalchemy/sqlalchemy/wiki/UsageRecipes) wiki
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sqlalchemy) tag
- SQLAlchemy [Google group](http://groups.google.com/group/sqlalchemy)
- [Gitter](https://gitter.im/sqlalchemy/community) chat

**Have a nice day!**

**Comment by zzzeek** [comment]
hey there -

there's nothing above to indicate connection behavior or what methodology you are using to observe that it is "recreating connections" each time, so I would advise turning on [pool logging](https://docs.sqlalchemy.org/en/14/core/engines.html#configuring-logging) using `echo_pool='debug'` and that will show you the decisions being made regarding connection pooling.

**Comment by sahil-khanna-instacart** [comment]
HI @zzzeek ,
Thanks for moving the issue to discussions and sharing instructions on logging. Logs seems to be returning the connection to the pool.

I was using the below mentioned snowflake authenticator and it keeps on asking me to authenticate with browser on every execution. 

I wanted to confirm that whether underline SQLalchemy QueuePool closes the connection when exiting out of below mentioned context manager or does it just return the connection to the pool?

```
Authenticator:
snowflake://***:@***/***?&authenticator=externalbrowser"
```
```
Context manager:
with engine.begin() as connection:
    connection.execute(...)
```

The confirmation would be help me triage the issue with my code.

Thanks for the help, it is very useful.

**Reply by zzzeek** [reply]
the pool closes the connection on return if it was in the overflow range.    the pool logging will show this.  it also closes the connection when it's invalidated or when it is older than the pool_recycle time.   check out the docs https://docs.sqlalchemy.org/en/14/core/pooling.html#sqlalchemy.pool.QueuePool

**Reply by sahil-khanna-instacart** [reply]
Thanks for the quick response and validate my understanding.
